packages/queued-request-controller/src/QueuedRequestController.ts

```diff
@@ -263,30 +286,48 @@ export class QueuedRequestController extends BaseController<
    */
   async #processNextBatch() {
     const firstRequest = this.#requestQueue.shift() as QueuedRequest;
-    this.#originOfCurrentBatch = firstRequest.origin;
-    this.#networkClientIdOfCurrentBatch = firstRequest.networkClientId;
-    const batch = [firstRequest.processRequest];
+    this.#originOfCurrentBatch = firstRequest.request.origin;
+    this.#networkClientIdOfCurrentBatch = firstRequest.request.networkClientId;
+    const batch = [firstRequest];
 
     // alternatively we could still batch by only origin but switch networks in batches by
     // adding the network clientId to the values in the batch array
     while (
-      this.#requestQueue[0]?.networkClientId ===
+      this.#requestQueue[0]?.request.networkClientId ===
         this.#networkClientIdOfCurrentBatch &&
-      this.#requestQueue[0]?.origin === this.#originOfCurrentBatch
+      this.#requestQueue[0]?.request.origin === this.#originOfCurrentBatch &&
+      !this.#canRequestSwitchNetworkWithoutApproval(
+        this.#requestQueue[0]?.request,
+      )
     ) {
       const nextEntry = this.#requestQueue.shift() as QueuedRequest;
-      batch.push(nextEntry.processRequest);
+      batch.push(nextEntry);
     }
-    // If globally selected network is different from origin selected network,
-    // switch network before processing batch
     let networkSwitchError: unknown;
     try {
-      await this.#switchNetworkIfNecessary(firstRequest.networkClientId);
+      const { request, processRequest, processedPromise } = firstRequest;
+      // If globally selected network is different from origin selected network,
+      // switch network before processing batch
+      await this.#switchNetworkIfNecessary(request.networkClientId);
+
+      // If the first request might switch the network, process it first
+      // and throw a error if the network now differs from that of the batch
+      if (this.#canRequestSwitchNetworkWithoutApproval(request)) {
+        batch.shift();
+        processRequest(networkSwitchError);
+        await processedPromise;
+        const { selectedNetworkClientId } = this.messagingSystem.call(
+          'NetworkController:getState',
+        );
+        if (this.#networkClientIdOfCurrentBatch !== selectedNetworkClientId) {
+          throw NetworkChangedError;
+        }
```

Comment:
was washing the dishes and realized that this also needs to call ` #flushQueueForOrigin`, since those requests are now invalid too

Reply:
perhaps the `this.#canRequestSwitchNetworkWithoutApproval(request)` should happen in the for loop and the batch doesn't have to end the batch with ` !this.#canRequestSwitchNetworkWithoutApproval(this.#requestQueue[0]?.request,`